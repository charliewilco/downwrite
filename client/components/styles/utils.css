.u-right {
  text-align: center;
}

.u-left {
  text-align: left;
}

.u-center {
  text-align: center;
}
