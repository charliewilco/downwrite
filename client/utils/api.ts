import {
  USER_ENDPOINT,
  POST_ENDPOINT,
  PREVIEW_ENDPOINT,
  SETTINGS_ENDPOINT,
  PASSWORD_ENDPOINT,
  AUTH_ENDPOINT,
  createURL
} from "./urls";
import * as Dwnxt from "downwrite";
import { __IS_DEV__, __IS_BROWSER__ } from "./dev";

import "isomorphic-fetch";
import "abortcontroller-polyfill/dist/abortcontroller-polyfill-only";

interface IOptions {
  token?: string;
  host?: string;
}

interface IUserResponse {
  userID: string;
  id_token: string;
  username: string;
  message?: string;
}

type APIResponse = Dwnxt.IPost | Dwnxt.IPostError;

type HeaderMethod = "GET" | "PUT" | "POST" | "DELETE";

const controller: AbortController = new AbortController();
const signal: AbortSignal = controller.signal;

export function abortFetching() {
  controller.abort();
}

/**
 * Creates header for Fetch request with Token
 * @param method
 * @param token
 */
export const createHeader = (
  method: HeaderMethod = "GET",
  token?: string
): RequestInit => {
  const Header = new Headers();
  if (token) {
    Header.set("Authorization", token);
  }
  Header.set("Content-Type", "application/json");

  return {
    method,
    headers: Header,
    mode: "cors",
    cache: "default",
    signal
  };
};

/**
 * Creates url string for both server and client based off header
 * @param url
 * @param hostname
 * @returns string
 */

export interface IAuthUserBody {
  user: string;
  password: string;
}

export async function authUser(
  { user, password }: IAuthUserBody,
  options?: IOptions
): Promise<any> {
  const url = createURL(AUTH_ENDPOINT, options.host);
  const auth = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ user, password })
  }).then(res => res.json());

  return auth;
}

export async function getUserDetails(options: IOptions): Promise<any> {
  const url = createURL(USER_ENDPOINT, options.host);
  const user = await fetch(url, createHeader("GET", options.token)).then(res =>
    res.json()
  );

  return user;
}

export async function updateSettings(
  body: {
    username: string;
    email: string;
  },
  options: IOptions
) {
  const url = createURL(SETTINGS_ENDPOINT, options.host);
  const settings = await fetch(url, {
    ...createHeader("POST", options.token),
    body: JSON.stringify(body)
  }).then(res => res.json());

  return settings;
}

export interface ICreateUserBody {
  username: string;
  email: string;
  password: string;
}

export async function createUser(
  { username, email, password }: ICreateUserBody,
  options?: IOptions
): Promise<IUserResponse> {
  const url = createURL(USER_ENDPOINT, options.host);

  const user = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ username, email, password })
  }).then(res => res.json());

  return user;
}

export async function updatePassword(body: any, options: IOptions): Promise<any> {
  const url = createURL(PASSWORD_ENDPOINT, options.host);
  const password = await fetch(url, {
    ...createHeader("POST", options.token),
    body: JSON.stringify(body)
  }).then(res => res.json());

  return password;
}

export async function findPreviewEntry(
  id: string,
  options?: IOptions
): Promise<null> {
  const url = createURL(PREVIEW_ENDPOINT, options.host);
  const entry = await fetch(`${url}/${id}`, {
    method: "GET",
    mode: "cors"
  }).then(res => res.json());

  return entry;
}

export async function getPost(id: string, options: IOptions): Promise<APIResponse> {
  const url = createURL(POST_ENDPOINT, options.host);
  const post = await fetch(`${url}/${id}`, createHeader("GET", options.token)).then(
    res => res.json()
  );

  return post;
}

export async function removePost(id: string, options: IOptions): Promise<Response> {
  const url = createURL(POST_ENDPOINT, options.host);
  const response = await fetch(
    `${url}/${id}`,
    createHeader("DELETE", options.token)
  );
  return response;
}

export async function getPosts(
  options: IOptions
): Promise<Dwnxt.IPost[] | Dwnxt.IPostError> {
  const url = createURL(POST_ENDPOINT, options.host);

  const entries: Dwnxt.IPost[] = await fetch(
    url,
    createHeader("GET", options.token)
  ).then(res => res.json());

  return entries;
}

export async function createPost(
  body: Dwnxt.IPostCreation,
  options: IOptions
): Promise<APIResponse> {
  const url = createURL(POST_ENDPOINT, options.host);
  const post = await fetch(url, {
    ...createHeader("POST", options.token),
    body: JSON.stringify(body)
  }).then(res => res.json());

  return post;
}

export async function updatePost(
  id: string,
  body: Dwnxt.IPost,
  options: IOptions
): Promise<Dwnxt.IPost | Dwnxt.IPostError> {
  const url = createURL(POST_ENDPOINT, options.host);
  const entry = await fetch(`${url}/${id}`, {
    ...createHeader("PUT", options.token),
    body: JSON.stringify(body)
  }).then(res => res.json());

  return entry;
}
