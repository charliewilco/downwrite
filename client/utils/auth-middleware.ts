import Router from "next/router";
import { NextContext } from "next";
import Cookies, { CookieGetOptions } from "universal-cookie";
import { __IS_BROWSER__ } from "./dev";

export interface ICookie {
  DW_TOKEN: string;
}

export function cookies<T>(
  context: NextContext,
  options?: CookieGetOptions
): T | {} {
  options = options || {};
  if (context.req) {
    // Server
    if (!context.req.headers) return {}; // for Static export feature of Next.js
    const cookies = new Cookies(context.req.headers.cookie);
    if (!cookies) return {};
    return cookies.getAll(options);
  } else {
    // Client
    const cookies = new Cookies();

    return cookies.getAll(options);
  }
}
// : Promise<string>
export const authMiddleware = (ctx: NextContext): string => {
  const cookie = cookies<ICookie>(ctx) as ICookie;
  const token = cookie.DW_TOKEN;

  if (ctx.req && !token) {
    ctx.res.writeHead(302, { Location: "/login" });
    ctx.res.end();
    return;
  }

  if (!token) {
    Router.push("/login");
  }

  return token;
};
