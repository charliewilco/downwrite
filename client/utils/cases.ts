import "jest";

interface Config {
  name?: string;
  only?: boolean;
  skip?: boolean;
}

// Based on: https://github.com/atlassian/jest-in-case

export function cases<T extends Config>(
  title: string,
  tester: (opts: T, done?: () => void) => Promise<any>,
  testCases: T[]
): void {
  let normalized: T[];

  if (Array.isArray(testCases)) {
    normalized = testCases;
  } else {
    const safeRef = testCases;
    normalized = Object.keys(testCases).map(name => {
      return Object.assign({}, { name }, safeRef[name]);
    });
  }

  describe(title, () => {
    normalized.forEach((testCase, index) => {
      let name = testCase.name || `case: ${index + 1}`;

      let testFn;
      if (testCase.only) {
        testFn = test.only;
      } else if (testCase.skip) {
        testFn = test.skip;
      } else {
        testFn = test;
      }

      let cb;
      if (tester.length > 1) {
        cb = (done: any) => tester(testCase, done);
      } else {
        cb = () => tester(testCase);
      }

      testFn(name, cb);
    });
  });
}
