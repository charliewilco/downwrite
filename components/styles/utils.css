.u-right {
  text-align: right;
}

.u-left {
  text-align: left;
}

.u-center {
  text-align: center;
}
