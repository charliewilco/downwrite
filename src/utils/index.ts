export * from "./api";
export * from "./classnames";
export * from "./dev";
export * from "./default-styles";
export * from "./markdown-template";
export * from "./sanitize";
export * from "./validations";
export * from "./urls";
