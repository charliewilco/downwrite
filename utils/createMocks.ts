import uuid from "uuid/v4";
import * as Draft from "draft-js";
import { IPost } from "downwrite";

export const createMockPost = (overides?: Partial<IPost>): IPost => {
  const post = Draft.EditorState.createWithContent(
    Draft.ContentState.createFromText("Hello")
  );
  const content = Draft.convertToRaw(post.getCurrentContent());
  const dateAdded = "2019-04-14T07:40:08.591Z";
  const id = uuid();

  return Object.assign(
    {},
    {
      title: "Something",
      id,
      dateAdded,
      content,
      public: false
    },
    overides || {}
  );
};

export const createMockPosts = (count: number): IPost[] => {
  return Array(count || 1)
    .fill(null)
    .map(() => createMockPost({ title: "Mocked Posts" }));
};

export const createEditorState = (content: string): Draft.EditorState => {
  const state = Draft.EditorState.createWithContent(
    Draft.ContentState.createFromText("Hello")
  );

  return state;
};
