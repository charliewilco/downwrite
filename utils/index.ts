export * from "./api";
export * from "./auth-middleware";
export * from "./classnames";
export * from "./dev";
export * from "./defaultStyles";
export * from "./local-drafts";
export * from "./markdownTemplate";
export * from "./sanitize";
export * from "./validations";
export * from "./urls";
export * from "./responseHandler";
export * from "./tokens";
