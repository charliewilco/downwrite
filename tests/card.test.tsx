import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { LinkProps } from "next/link";
import { fireEvent, render } from "@testing-library/react";
import Card from "../components/card";
import { createMockPost } from "../utils/createMocks";

const title = "Starting Again";
const post = createMockPost({ title, id: "4444" });
const mockDelete = jest.fn();

jest.mock("next/link", () => {
  return jest.fn((props: React.PropsWithChildren<LinkProps>) => (
    <>{props.children}</>
  ));
});

const { container, getByTestId } = render(
  <Card public={false} {...post} onDelete={mockDelete} />
);

// NOTE: test broken by upgrading @testing-library
xdescribe("<Card />", () => {
  it("contains snippet from content", () => {
    const snippet = getByTestId("CARD_EXCERPT").textContent;
    expect(snippet.length).toBeLessThanOrEqual(90);
  });

  it("contains a title", () => {
    expect(post.title).toBe(title);
    expect(getByTestId("CARD_TITLE")).toHaveTextContent(title);
  });

  it("contains delete button", () => {
    expect(getByTestId("CARD_DELETE_BUTTON")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should fire a delete", () => {
    fireEvent.click(getByTestId("CARD_DELETE_BUTTON"));

    expect(mockDelete).toHaveBeenCalled();
  });
});
