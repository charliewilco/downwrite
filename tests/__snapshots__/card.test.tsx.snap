// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> matches snapshot 1`] = `
<div
  class="Sheet Card"
  data-testid="CARD"
>
  <header
    class="CardHeader"
  >
    <h2
      class="CardTitle"
      data-testid="CARD_TITLE"
    >
      <a>
        Starting Again
      </a>
    </h2>
    <small
      class="CardMeta"
    >
      added 
      less than a minute
       ago
    </small>
  </header>
  <footer
    class="CardTray"
  >
    <div
      class="links"
      data-testid="CARD_EXCERPT"
    >
      <a
        style="margin-right: 8px;"
      >
        Edit
      </a>
    </div>
    <button
      class="CardDeleteButton"
      data-testid="CARD_DELETE_BUTTON"
    >
      Delete
    </button>
  </footer>
</div>
`;
